alpenhorn/update.py

```diff
@@ -0,0 +1,466 @@
+    #     return
+
+    # Make sure this node is usable.
+    if not node.mounted:
+        log.debug("Skipping unmounted node \"%s\"." % node.name)
+        return
+    if node.suspect:
+        log.debug("Skipping suspected node \"%s\"." % node.name)
+
+    log.info("Updating node \"%s\"." % (node.name))
+
+    # Check and update the amount of free space then reload the instance for use
+    # in later routines
+    update_node_free_space(node)
+
+    # Check the integrity of any questionable files (has_file=M)
+    update_node_integrity(node)
+
+    # Delete any upwanted files to cleanup space
+    update_node_delete(node)
+
+    # Process any regular transfers requests onto this node
+    update_node_requests(node)
+
+    # TODO: bring back HPSS support
+    # Process any tranfers out of HPSS onto this node
+    # update_node_hpss_outbound(node)
+
+
+def update_node_free_space(node):
+    """Calculate the free space on the node and update the database with it."""
+
+    # Check with the OS how much free space there is
+    x = os.statvfs(node.root)
+    avail_gb = float(x.f_bavail) * x.f_bsize / 2**30.0
+
+    # Update the DB with the free space. Save only the dirty fields to ensure we
+    # don't clobber changes made manually to the database
+    node.avail_gb = avail_gb
+    node.avail_gb_last_checked = datetime.datetime.now()
+    node.save(only=node.dirty_fields)
+
+    log.info("Node \"%s\" has %.2f GB available." % (node.name, avail_gb))
+
+
+def update_node_integrity(node):
+    """Check the integrity of file copies on the node."""
+
+    # Find suspect file copies in the database
+    fcopy_query = ar.ArchiveFileCopy.select().where(
+        ar.ArchiveFileCopy.node == node,
+        ar.ArchiveFileCopy.has_file == 'M').limit(25)
+
+    # Loop over these file copies and check their md5sum
+    for fcopy in fcopy_query:
+        fullpath = "%s/%s/%s" % (node.root, fcopy.file.acq.name, fcopy.file.name)
+        log.info("Checking file \"%s\" on node \"%s\"." % (fullpath, node.name))
+
+        # If the file exists calculate its md5sum and check against the DB
+        if os.path.exists(fullpath):
+            if util.md5sum_file(fullpath) == fcopy.file.md5sum:
+                log.info("File is A-OK!")
+                fcopy.has_file = 'Y'
+            else:
+                log.error("File is corrupted!")
+                fcopy.has_file = 'X'
+        else:
+            log.error("File does not exist!")
+            fcopy.has_file = 'N'
+
+        # Update the copy status
+        log.info("Updating file copy status [id=%i]." % fcopy.id)
+        fcopy.save()
+
+
+def update_node_delete(node):
+    """Process this node for files to delete."""
+
+    # If we are below the minimum available size, we should consider all files
+    # not explicitly wanted (wants_file != 'Y') as candidates for deletion,
+    # otherwise only those explicitly marked (wants_file == 'N')
+    # Also do not clean on archive type nodes.
```

Comment:
Yeah. That's unclear. We allow cleaning on archive nodes when there are enough copies elsewhere and they explicitly marked as `wants_file='N'`. I'll clarify the comment.